search_similar_entities/facetedSearchEndpoint.rb

```diff
@@ -0,0 +1,161 @@
+#QuerySimilar Endpoint
+
+#Builds complex query for"Query Similar" functionality; returns a link to the search page, with a preseeded query looking for all entities of type testEntity that share certain attributes with the current "testEntity" entity, or alternatively, the same query but with preselected facets
+
+# Verify params are OK
+raise "Missing required parameter: id" unless params[:id]
+raise "Missing required parameter: mode - Values: [queryString|queryStringWithFacets]" unless params[:mode]
+raise "Acceptable values for parameter mode: [queryString|queryStringWithFacets]" unless "queryString|queryStringWithFacets|".include? params["mode"] + '|'
+thisId = params[:id]
+return_mode = params[:mode]
+
+## Utility Methods
+
+# Utility method to return the different "leaf" values found in the hierarchical attribute 
+# 'hierachicalAttr1' belonging
+# to a testEntity, that begin with level1value|level2value
+def getLeafValues(testEntity, level1value, level2value)
+  return testEntity['hierarchicalAttr1'].select { |a| a.start_with?(level1value + '|' + level2value) }.uniq
+end
+
+# Utility method that adds n and k elements to a hash
+def addNKElementsToHash(theHash, nObj, kObj)
+  unless nObj.nil? or nObj.empty?
+    theHash['n'] = nObj
+  end
+  unless kObj.nil? or kObj.empty?
+    theHash['k'] = kObj
+  end
+  return theHash
+end
+  
+
+# Utility method to return the data structure that goes in the queryURL (as JSON) 
+# to make sure a given refinement widget is selected for a given value
+def getSimpleRefWidgetDataStruct(widgetName,xPath,logic,theValue)
+  
+  theValueHash = {"n" => theValue}
+  tmpHash = {"id" => widgetName, "logic" => logic, "s" => [theValueHash]}
+  unless xPath.empty?
+    tmpHash['xPath'] = xPath
+  end
+  return tmpHash
+  
+end 
+
+# Utility method to return the data structure that goes in the queryURL (as JSON) 
+# to make sure a given refinement widget (that has hierarchical values) is selected for a given value
+def getComplexRefWidgetDataStruct(widgetName,xPath,logic,theValue,separator)
+  
+  tmpHash = {"id" => widgetName, "logic" => logic, "s" => theValue}
```

Comment:
What is `tmpHash` for?  With a more descriptive variable name the code becomes part of the tutorial.